published/external/xdpapi.h

```diff
@@ -156,7 +228,269 @@ XdpUnloadApi(
     FreeLibrary(XdpHandle);
 }
 
-#endif // !defined(_KERNEL_MODE)
+#endif // defined(_KERNEL_MODE)
+
+#if defined(_KERNEL_MODE)
+
+#include <netioddk.h>
+#include <xdp/guid.h>
+
+#define POOLTAG_NMR_CLIENT      'cNdX'   // XdNc
+
+typedef struct _XDP_LOAD_CONTEXT *XDP_LOAD_API_CONTEXT;
+
+typedef
+VOID
+XDP_API_ATTACH_FN(
+    _In_ VOID *ClientContext
+    );
+
+typedef
+VOID
+XDP_API_DETACH_FN(
+    _In_ VOID *ClientContext
+    );
+
+typedef struct _XDP_API_CLIENT {
+    ERESOURCE Resource;
+
+    //
+    // NMR client registration.
+    //
+    NPI_CLIENT_CHARACTERISTICS NpiClientCharacteristics;
+    NPI_MODULEID ModuleId;
+    HANDLE NmrClientHandle;
+
+    //
+    // XDPAPI client.
+    //
+    VOID *Context;
+    XDP_API_ATTACH_FN *Attach;
+    XDP_API_DETACH_FN *Detach;
+    const XDP_API_CLIENT_DISPATCH *XdpApiClientDispatch;
+
+    //
+    // XDPAPI provider.
+    //
+    HANDLE BindingHandle;
+    XDP_API_PROVIDER_DISPATCH *XdpApiProviderDispatch;
+    XDP_API_PROVIDER_BINDING_CONTEXT *XdpApiProviderContext;
+} XDP_API_CLIENT;
+
+inline
+NTSTATUS
+XdpNmrClientAttachProvider(
+    _In_ HANDLE _NmrBindingHandle,
+    _In_ VOID *_ClientContext,
+    _In_ const NPI_REGISTRATION_INSTANCE *_ProviderRegistrationInstance
+    )
+{
+    XDP_API_CLIENT *_Client = (XDP_API_CLIENT *)_ClientContext;
+    NTSTATUS _Status;
+    VOID *_ProviderBindingContext;
+    const VOID *_ProviderBindingDispatch;
+
+    //
+    // The NMR client allows at most one active binding at a time, but defers
+    // all remaining binding logic to the NMR provider.
+    //
+
+    ExEnterCriticalRegionAndAcquireResourceExclusive(&_Client->Resource);
+
+    if (_Client->BindingHandle != NULL) {
+        _Status = STATUS_DEVICE_NOT_READY;
+    } else if (_ProviderRegistrationInstance->Number != XDP_API_VERSION_1) {
+        _Status = STATUS_NOINTERFACE;
+    } else {
+        _Status =
+            NmrClientAttachProvider(
+                _NmrBindingHandle, _Client, &_Client->XdpApiClientDispatch,
+                &_ProviderBindingContext, &_ProviderBindingDispatch);
+
+        if (NT_SUCCESS(_Status)) {
+            _Client->BindingHandle = _NmrBindingHandle;
+            _Client->XdpApiProviderDispatch = (XDP_API_PROVIDER_DISPATCH *)_ProviderBindingDispatch;
+            _Client->XdpApiProviderContext = _ProviderBindingContext;
+        }
+    }
+
+    ExReleaseResourceAndLeaveCriticalRegion(&_Client->Resource);
+
+    if (NT_SUCCESS(_Status) && _Client->Attach != NULL) {
+        _Client->Attach(_Client->Context);
+    }
+
+    return _Status;
+}
+
+inline
+NTSTATUS
+XdpNmrClientDetachProvider(
+    _In_ VOID *_ClientBindingContext
+    )
+{
+    XDP_API_CLIENT *_Client = (XDP_API_CLIENT *)_ClientBindingContext;
+
+    if (_Client->Detach != NULL) {
```

Comment:
if the client doesn't provide a detach routine, then xdp can simply be unloaded from underneath the client? i think there needs to be more synchronization provided in a standardized way, similar to how the IO manager prevents the driver being unloaded while any handles are open or IOCTLs in flight

Reply:
Is there any reference I can learn what you mean?
And I cannot answer quickly.
@nigriMSFT could you help?

Reply:
I think we should just require the Attach/Detach routines from the client. It doesn't make sense for them to not be in the loop on cleanup.